server/src-lib/Hasura/Eventing/EventTrigger.hs

```diff
@@ -1,6 +1,50 @@
+{-|
+= Event Triggers
```

Comment:
This comment is helpful, but I think it goes too far into minor details of the implementation. It’s described very operationally, as if it were pseudocode for an algorithm: “if this happens, we do A, else we do B.” This makes the comment a little harder to follow (the reader has to keep track of all these details in their head!) and also much more likely to rapidly go out of date.

Instead of describing eventing like a step-by-step algorithm, it would be more useful to talk about the different concepts at play and describe their role in the overall system, without going into how exactly they’re implemented. Doing so will naturally explain the *why* in addition to the *what*. For example, you might start by saying:

> Event triggers are like ordinary SQL triggers, except instead of calling a SQL procedure, they call a webhook. This means event delivery involves coordination between both the database and `graphql-engine`: only the SQL database knows when the events should fire, but only `graphql-engine` knows how to actually deliver them.
>
> Therefore, event triggers are implemented in two parts:
>
>   1. Every event trigger is backed by a bona fide SQL trigger. When the SQL trigger fires, it simply creates a new record in the `hdb_catalog.event_log` table and returns.
>
>   2. Concurrently, a thread in `graphql-engine` monitors the `event_log` table for new records. When it sees a new event, it uses the information stored in the row to carry out the remainder of the delivery process.

From that alone, it’s immediately clear (1) what event triggers are, (2) what problem they solve for our users, and (3) what the high-level implementation strategy is. Once you’ve established that, you can go into a little more detail about the implementation, if you’d like:

> The creation and deletion of the SQL trigger itself is managed by the metadata DDL APIs (see Hasura.RQL.DDL.EventTrigger), so this module focuses on event delivery. `processEventQueue` is the high-level entry point: it loops forever, scanning for new events. When it finds one, it dispatches a worker thread to asynchronously call the webhook.
>
> Most of the subtleties involve guaranteeing *reliable* delivery of events: we guarantee that every event will be delivered at least once, even if `graphql-engine` crashes. This means we have to record the state of each event in the database, and we have to retry failed requests at a regular (user-configurable) interval.

The other details probably shouldn’t go in the module-level comment. If they need to go anywhere, they should go in other comments elsewhere in the module. Maybe it makes sense to put them on a particular function’s documentation, or maybe it makes sense to put them [in a `Note`](https://github.com/ecthiender/graphql-engine/blob/hs-style-guide/server/STYLE.md#out-of-line-notes) and to reference them from the relevant place, but that’s your call to make.

Reply:
While documenting the above, I felt that maybe it was an overkill with all the implementation details. Your suggestion makes much more sense to add it as the module level documentation. I have moved the implementation details in a `Note` and I have referenced it from the appropriate places in the code.